crates/services/src/seqlock.rs

```diff
@@ -0,0 +1,158 @@
+//! A simple implementation of a sequential lock.
+//! More details: <https://docs.kernel.org/locking/seqlock.html>
+
+use std::{
+    cell::UnsafeCell,
+    sync::atomic::{
+        AtomicU64,
+        Ordering,
+    },
+};
+
+/// A simple implementation of a sequential lock.
+/// some usage of unsafe, T must be Copy
+#[derive(Debug)]
+pub struct SeqLock<T> {
+    sequence: AtomicU64,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send> Sync for SeqLock<T> {}
+
+/// The writer handle for the `SeqLock`.
+/// Only one writer exists for a `SeqLock`.
+#[derive(Debug)]
+pub struct SeqLockWriter<T> {
+    lock: std::sync::Arc<SeqLock<T>>,
+}
+
+impl<T> SeqLockWriter<T> {
+    /// Modifies the data within the lock.
+    pub fn write<F>(&self, f: F)
+    where
+        F: FnOnce(&mut T),
+    {
+        let lock = &self.lock;
+
+        // Indicate that a write operation is starting.
+        lock.sequence.fetch_add(1, Ordering::Release);
+
+        // Modify the data.
+        unsafe {
```

Comment:
Needs `// Safety` comment to explain why the `unsafe´ is acceptable.

Reply:
addressed in 72e09f9beb